actions/ulog_ingestion/src/ulog_ingestion/__main__.py

```diff
@@ -0,0 +1,304 @@
+import argparse
+import os
+from typing import List
+from pyulog.core import ULog
+import logging
+import sys
+import pathlib
+import time
+from concurrent.futures import ProcessPoolExecutor
+
+from roboto.association import (
+    Association,
+    AssociationType,
+)
+from roboto.domain import actions, datasets, files, topics
+from roboto.http import (
+    HttpClient,
+    SigV4AuthDecorator,
+)
+from roboto.transactions import TransactionManager
+
+import ulog_ingestion.utils as utils
+
+log = logging.getLogger("Ingesting ULog files to Roboto")
+
+
+def load_env_var(env_var: actions.InvocationEnvVar) -> str:
+    """
+    Load an environment variable, and exit if it is not found.
+
+    Args:
+    - env_var: The environment variable to load.
+
+    Returns:
+    - The value of the environment variable.
+    """
+    value = os.getenv(env_var.value)
+    if not value:
+        log.error("Missing required ENV var: '%s'", env_var)
+        sys.exit(1)
+    return value
+
+
+def setup_env():
+    """
+    Set up the environment for the action.
+
+    Returns:
+    - A tuple containing the organization ID, input directory, output directory, topic delegate, and dataset.
+    """
+    roboto_service_url = load_env_var(actions.InvocationEnvVar.RobotoServiceUrl)
+    org_id = load_env_var(actions.InvocationEnvVar.OrgId)
+    invocation_id = load_env_var(actions.InvocationEnvVar.InvocationId)
+    input_dir = load_env_var(actions.InvocationEnvVar.InputDir)
+    output_dir = load_env_var(actions.InvocationEnvVar.OutputDir)
+
+    http_client = HttpClient(default_auth=SigV4AuthDecorator("execute-api"))
+
+    topic_delegate = topics.TopicHttpDelegate(
+        roboto_service_base_url=roboto_service_url, http_client=http_client
+    )
+
+    invocation = actions.Invocation.from_id(
+        invocation_id,
+        invocation_delegate=actions.InvocationHttpDelegate(
+            roboto_service_base_url=roboto_service_url, http_client=http_client
+        ),
+        org_id=org_id,
+    )
+    dataset = datasets.Dataset.from_id(
+        invocation.data_source.data_source_id,
+        datasets.DatasetHttpDelegate(
+            roboto_service_base_url=roboto_service_url, http_client=http_client
+        ),
+        files.FileClientDelegate(
+            roboto_service_base_url=roboto_service_url, http_client=http_client
+        ),
+        transaction_manager=TransactionManager(
+            roboto_service_base_url=roboto_service_url, http_client=http_client
+        ),
+    )
+
+    return org_id, input_dir, output_dir, topic_delegate, dataset
+
+
+def process_wrapper(args):
+    """
+    Wrapper function for parallel processing.
+    """
+    return process_data(*args)
+
+
+def process_data(
```

Comment:
To get around questions of whether certain of these objects are process-safe or not, I would consider avoiding passing these in as arguments and instead initializing them within this function. That includes things like our delegates, but also the `ULog` object (are numpy arrays process-safe?). I'd also consider not pre-initializing the `schema_registry_dict` or `schema_checksum_dict` and just doing that in each individual process.

So, doing less (a lot less) in the top-level `ingest_ulog` function and more here.

Reply:
That makes sense - I refactored the code accordingly.  I moved most of the processing to the process_data function.

Reply:
Looks good!